forcetree.c

```diff
@@ -684,51 +687,88 @@ add_particle_moment_to_node(struct NODE * pnode, int i)
- *  flags whether the node contains any particles with lower softening
- *  than that.
+ *  This function is called recursively using openmp tasks.
+ *  We spawn a new task for a fixed number of levels of the tree.
  *
- *  The function also concatenates the NextNode linked lists. The return value
- *  and argument tail is the current tail of the NextNode linked list.
  */
-
-int
-force_update_node_recursive(int no, int sib, int tail, const struct TreeBuilder tb)
+static int
+force_update_node_recursive(int no, int sib, int level, const struct TreeBuilder tb)
 {
-    /*Set NextNode for this node*/
-    if(tail < tb.firstnode && tail >= 0 && force_get_next_node(tail, tb) != -1) {
-        endrun(2,"Particle %d with tail %d already has tail set: %d\n",no, tail, force_get_next_node(tail, tb));
-    }
-    tail = force_set_next_node(tail, no, tb);
-
-    /* For particles and pseudo particles we have nothing to update; */
-    /* But the new tail is the last particle in the linked list. */
-    if(no < tb.firstnode || no >= tb.lastnode) {
-        int next = no;
-        while(next != -1) {
-            no = next;
-            next = force_get_next_node(next, tb);
-        }
-        return no;
-    }
+    /*Last value of tails is the return value of this function*/
+    int j, suns[8], tails[8];
+
+    /* Only start spawning new tasks when we are past the top level nodes,
+     * to prevent us being overwhelmed by very short tasks pointing to pseudoparticles*/
+    if(level < 0 && Nodes[no].f.TopLevel && !Nodes[no].f.InternalTopLevel)
+        level = 1;
 
-    int j, suns[8];
+    /*Count how many internal children we have,
+     *so we can keep track of how many tasks we started*/
+    int chldcnt=0;
     /* this "backup" is necessary because the nextnode
      * entry will overwrite one element (union!) */
-    for(j = 0; j < 8; j++)
+    for(j = 0; j < 8; j++) {
         suns[j] = Nodes[no].u.suns[j];
+        if(suns[j] >= tb.firstnode && suns[j] < tb.lastnode)
+            chldcnt++;
+    }
+
+    /*First do the children*/
+    for(j = 0; j < 8; j++)
+    {
+        int p = suns[j];
+        /*Empty slot*/
+        if(p < 0)
+            continue;
+        /* For particles and pseudo particles we have nothing to update; */
+        /* But the new tail is the last particle in the linked list. */
+        if(p < tb.firstnode || p >= tb.lastnode) {
+            int next = p;
+            while(next != -1) {
+                p = next;
+                next = force_get_next_node(next, tb);
+            }
+            tails[j] = p;
+        }
+        else {
+            const int nextsib = force_get_sibling(sib, j, suns);
+            /*Don't spawn a new task if we only have one child,
+             *or if we are deep enough that we already spawned a lot.
+             Note: final clause is much slower for some reason. */
+            if(chldcnt > 1 && level >= 0 && level < 513) {
+                #pragma omp task shared(tails)
```

Comment:
Didn't the final clause work?

Reply:
I think it is because the compiler is still emitting the code to do the book keeping of tasks. I am curious if mergeable helps (some say yes); but that's not very important. The current one is already very clean.

Reply:
What you say about final sounds plausible, but note that it slows things down even if the conditional is unchanged (so the final clause is always false). This suggests some kind of locking to me. mergeable is also slower (I just tested it). I hadn't seen it before, but it seems to give the compiler the option to not generate a task?  

The other thing I noticed was that one cannot have an array element in the shared clause. So manually unrolling the loop might help a little with contention. But I did not do that, because it is ugly.